### ISSUE TITLE
Review the creation and use of a dart_context

### ISSUE THREAD

**Issue by nshahan** [dart-lang/rules_dart#65]
Needs some documentation explaining lib_root, label.workspace_root and label.package and what they are used for.

There is some confusing usage that should be cleaned up when we clearly define the usage of these fields.
https://github.com/dart-lang/rules_dart/blob/master/dart/build_rules/internal/common.bzl#L109 
https://github.com/dart-lang/rules_dart/blob/master/dart/build_rules/internal/common.bzl#L210

**Comment by cbracken** [comment]
`lib_root`: the directory that specifies the location of library srcs relative to the *Bazel package*. This is typically the `lib` directory of a *Dart package* that follows Pub layout conventions.

`workspace_root`: This is a Bazel concept, not Dart specific. It's a user-specifiable directory (configured via a `workspace()` call in the WORKSPACE file). It's optional or could be set to an arbitrary value, say `google3`.

`Label.package`: This is a Bazel concept, not Dart specific. This specifies the *Bazel package* to which a Label refers.

Bazel package: the directory in which a BUILD file is located.

**Comment by nshahan** [comment]
To clarify, in this example the `lib_root` should be "bar/lib/"? 

The path to the lib folder relative to the location of the BUILD file that defines the `dart_library` in question?

**Comment by cbracken** [comment]
Correct; `lib_root` is intended to be a Bazel-package-relative (i.e. BUILD-file-relative) path to the directory containing the library srcs for the dart_library in question.
